### PR TITLE
Detect broadcast+random and bail out

### PR DESCRIPTION
**Pull request by bertmaher**
This is based on the fallback interpreter PR, #247.  If we find a TE::group that contains broadcasts and random, bail out on the whole thing.  This is what the current graph fuser does.

I don't exactly love the inelegance of having broadcastShapes return a pair.

I considered a few more complicated things:

- Do a backward dataflow analysis while building a TE::group to avoid absorbing a `rand` into a region that involves a broadcast.  That would add a fair bit of complexity to tensorexpr_fuser.
- Do dataflow analysis while scheduling, and do not inline an expression containing rand() into a broadcast.  That also adds a fair bit of complexity, and would require us to support non-inline computations, which means supporting `Allocate` on all backends.  I'm wary of doing `Allocate` at this stage.

Basically this is the easiest thing that works, which seems like the way to go unless there's evidence that broadcasted randoms are important.

### CODE REVIEW

**Review by zheng-xq — COMMENTED** [review]
Could you rebase your change? The change looks very messy now.

**Comment by bertmaher** [comment]
Rebased.  Note that this PR includes changes from #247, which I think is a good idea independently of this PR :)

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 1]
```diff
@@ -1189,17 +1218,43 @@ TensorExprKernel::TensorExprKernel(const Graph& subgraph) {
...
     tensor_outputs_.emplace_back(tensors_.at(output->unique()));
     tensors_.erase(output->unique());
   }
 }
 
+TensorExprKernel::TensorExprKernel(const std::shared_ptr<Graph>& subgraph)
+    : graph_(subgraph), code_(subgraph) {
+  try {
+    compile();
+  } catch (...) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp]: We probably need to catch a specific exception, not everything.

File: torch/csrc/jit/tensorexpr/kernel.h [hunk 2]
```diff
@@ -55,6 +60,10 @@ class TensorExprKernel {
     kCudaCodeGen,
   };
 
+  void compile();
+
+  void runKernel(Stack& stack);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.h]: Nit: we probably need to start adding some comments to class methods/members.

**Review by ZolotukhinM — APPROVED** [review]
Both fallback and broadcast-detection parts look good to me. I think eventually we'd need to detect the broadcasting earlier (when we try to inline), but it's good to have this safety fallback mechanism in place anyway.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 3]
```diff
@@ -1189,17 +1218,43 @@ TensorExprKernel::TensorExprKernel(const Graph& subgraph) {
         }
       }
     }
+    if (hasRandom_ && hasBroadcast_) {
+      throw std::runtime_error(
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1346]: Within reason, I'd like to use a return value to indicate that the fusion failed, instead of through an exception. 

An exception within a debugger makes it hard if someone is trying to get to the real exceptions. Also it is slow enough, if we are autotuning through many possibilities, it might make the whole thing thousands of times slower.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 1 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: In general I agree but I'm following the current fuser on this specific point - if anything goes wrong during compilation I want to be able to hand the reins to the interpreter.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 3 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1346]: I'd like to come back to this at some point.  I want the try/catch for general robustness purposes; I can see using an error code here instead but I want to think through the error codes properly.
